docs: replace com.vaadin.hilla.Nonnull usages (#4155)
vaadin/docs: articles/hilla/lit/reference/type-nullability.adoc

- A collection accepts `null`, unless the collection item type is primitive.
- A map accepts `null`, unless the collection item type is primitive.

> Any of these nullable types can be made non-nullable by applying a `@NonNull` annotation. Vaadin recommends using the `@org.jspecify.annotations.NonNull`. However, the list of all supported annotations is as follows:

Reviewer: ```suggestion
Any of these nullable types can be made non-nullable by applying a `@NonNull` annotation. Vaadin recommends using `@org.jspecify.annotations.NonNull`. The full list of supported annotations is as follows:
```

- A collection accepts `null`, unless the collection item type is primitive.
- A map accepts `null`, unless the collection item type is primitive.

Any of these nullable types can be made non-nullable by applying a `@NonNull` annotation.

> While the usage of many annotations that has the name `nonnull` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.NonNull`.

Reviewer: Not true at the moment: we don't support annotations by name pattern, instead we have a list of class names.

Reviewer: See https://github.com/vaadin/hilla/blob/4cfcab36e989583e5be2101205b7338c148dd0e8/packages/java/parser-jvm-plugin-nonnull/src/main/java/com/vaadin/hilla/parser/plugins/nonnull/NonnullPluginConfig.java#L66-L72

- A collection accepts `null`, unless the collection item type is primitive.
- A map accepts `null`, unless the collection item type is primitive.

Any of these nullable types can be made non-nullable by applying a `@NonNull` annotation.

> While using any annotation that has the name `nonnull` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.NonNull`. Below are some examples of other supported `nonnull` annotations:

Reviewer 2: ```suggestion
While the usage of any annotation that has the name `nonnull` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.NonNull`. Below are some examples of other supported `nonnull` annotations:
```

> While Hilla supports using any annotation that has the name `nonnull` (case-insensitive), Vaadin recommends using the `org.jspecify.annotations.NonNull`. Below are some examples of other supported `nonnull` annotations:

Reviewer 2: ```suggestion
While using any annotation that has the name `nonnull` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.NonNull`. Below are some examples of other supported `nonnull` annotations:
```

> 

Reviewer 2: I addressed this a while ago, so it could get merged IMO, unless there's something more to address.

== `@NonNullApi`

Along with `@NonNull` annotations, you could also use package-level `@NonNullApi` annotations. It would make all the nullable types in a package non-nullable by default. All nested types -- `List` and `Map` items, etc. -- are also affected.

By default, the following annotation is supported: `org.springframework.lang.NonNullApi`.

> To make any type nullable, you must add `@Nullable` annotation to it. While the usage of many annotations that has the name `nullable` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.Nullable`.

Reviewer: As above: https://github.com/vaadin/hilla/blob/4cfcab36e989583e5be2101205b7338c148dd0e8/packages/java/parser-jvm-plugin-nonnull/src/main/java/com/vaadin/hilla/parser/plugins/nonnull/NonnullPluginConfig.java#L58-L62

Reviewer 2: OK, any suggestions? The previous review asked to remove the bullet list, should we bring it back?

Reviewer 2: I think the motivation for supporting a wide range of `@Nullable` annotations was for easier adoption in the already existing code bases. This is still a valid argument IMO. If the current functionality doesn't match the intention behind this design, we should fix that instead. 

For now, I'd write something like this that clearly reflects the supported ones:
```
To make any type nullable, you must add `@Nullable` annotation to it. The following are the supported annotations:
- `org.jspecify.annotations.Nullable` (Recommended)
- `jakarta.annotation.Nullable`
- `org.springframework.lang.Nullable`
- `com.vaadin.hilla.Nullable` (Deprecated)

Vaadin strongly recommends using `org.jspecify.annotations.Nullable`. Use the other supported annotations only if there is a compatibility issue or any other limitation.
```
WDYT?

Reviewer 3: @taefi keeping the list is a good idea. Let's rephrase to make jspecify one mentioned first, and then continue with the list of other supported annotations.

Reviewer 2: @platosha done.

== `@NonNullApi`

Along with `@NonNull` annotations, you could also use package-level `@NonNullApi` annotations. It would make all the nullable types in a package non-nullable by default. All nested types -- `List` and `Map` items, etc. -- are also affected.

By default, the following annotation is supported: `org.springframework.lang.NonNullApi`.

> To make any type nullable, you must add `@Nullable` annotation to it. Vaadin recommends using the `@org.jspecify.annotations.Nullable`. However, the list of all supported annotations is as follows:

Reviewer: ```suggestion
To make any type nullable, you must add a `@Nullable` annotation to it. Vaadin recommends using `@org.jspecify.annotations.Nullable`. The full list of supported annotations is as follows:
```

== `@NonNullApi`

Along with `@NonNull` annotations, you could also use package-level `@NonNullApi` annotations. It would make all the nullable types in a package non-nullable by default. All nested types -- `List` and `Map` items, etc. -- are also affected.

By default, the following annotation is supported: `org.springframework.lang.NonNullApi`.

> To make any type nullable, you must add `@Nullable` annotation to it. While any annotation that has the name `nullable` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.Nullable`.

Reviewer 2: ```suggestion
To make any type nullable, you must add `@Nullable` annotation to it. While the usage of any annotation that has the name `nullable` (case-insensitive) is supported, Vaadin recommends using the `org.jspecify.annotations.Nullable`.
```
